api/proto/teleport/legacy/types/types.proto

```diff
@@ -1892,6 +1897,145 @@ message ProvisionTokenSpecV2Oracle {
+    // reuse of this token provided the client has access to the keypair. This
+    // may be useful in certain cases - like use in unsupported CI/CD providers
+    // - but cannot offer the same security assurances and should be used with
+    // care.
+    bool Insecure = 4 [(gogoproto.jsontag) = "insecure"];
+  }
+
+  // Onboarding contains parameters related to initial joining and keypair
+  // registration.
+  OnboardingSpec Onboarding = 1 [(gogoproto.jsontag) = "onboarding"];
+
+  // Joining contains parameters related to all join attempts.
+  JoiningSpec Joining = 2 [(gogoproto.jsontag) = "joining"];
+
+  // RotateOnNextRenewal is a flag that indicates clients should perform a
+  // keypair rotation on the next refresh or rejoin attempt. This flag will
+  // automatically be set to `false` after a successful rotation.
+  bool RotateOnNextRenewal = 3 [(gogoproto.jsontag) = "rotate_on_next_renewal"];
+}
+
+// ProvisionTokenStatusV2 contains status information about a particular
+// ProvisionTokenV2. These fields are generally not writable by end users.
+message ProvisionTokenStatusV2 {
+  // BoundKeypair contains status information related to bound-keypair type
+  // tokens.
+  ProvisionTokenStatusV2BoundKeypair BoundKeypair = 1 [(gogoproto.jsontag) = "bound_keypair,omitempty"];
+}
+
+message ProvisionTokenStatusV2BoundKeypair {
+  // InitialJoinSecret contains a secret value that may be used for public key
+  // registration during the initial join process. If
+  // `.spec.bound_keypair.onboarding.initial_public_key` is set, †his field will
+  // remain empty. Otherwise, if
+  // `.spec.bound_keypair.onboarding.initial_join_secret` is set, that value
+  // will be copied here. If that field is unset, a value will be randomly
+  // generated.
+  string InitialJoinSecret = 1 [(gogoproto.jsontag) = "initial_join_secret"];
+
+  // BoundPublicKey contains the currently bound public key. If
+  // `.spec.bound_keypair.onboarding.initial_public_key` is set, that value will
+  // be copied here on creation, otherwise it will be populated as part of
+  // public key registration. This value will be updated over time if keypair
+  // rotation takes place, and will always reflect the currently valid public
+  // key. This value is written in SSH public key format, including algorithm.
+  string BoundPublicKey = 2 [(gogoproto.jsontag) = "bound_public_key"];
+
+  // BoundBotInstanceID is the ID of the currently associated bot instance. A
+  // new bot instance is issued on each join; the new bot instance will
+  // have a `previous_bot_instance` set to this value, if any.
+  string BoundBotInstanceID = 3 [(gogoproto.jsontag) = "bound_bot_instance_id"];
+
+  // RemainingJoins is a count of the total number of additional joins that can
+  // be performed using this token. If `.spec.joining.total_joins` is
+  // incremented, this value will be incremented by the same value.
+  uint32 RemainingJoins = 4 [(gogoproto.jsontag) = "remaining_joins"];
```

Comment:
Is this field needed? It feels like it is a bit redundant and adds a bit of complexity.

Does this count the number of remaining join attempts or successful joins? Would this essentially be the same value as `total_joins` - `join_count`?

Can `total_joins` be decremented? What happens if `total_joins` is decremented by more than the number of `remaining_joins` and then incremented again?

Might be confusing to the user if the `0` value can indicate that there are no more remaining joins or there are unlimited joins.

Reply:
Huh, I think you're right. That removes an awkward calculated field and makes all the counters make sense regardless of whether or not `unlimited` is enabled.

For context, `remaining_joins` was originally the only field, and the zero value was just going to be a bit confusing if `unlimited` was set; `join_count` was only added later for informational purposes. But... I think you're right, using that field for everything is much simpler.

I'll make the change. Good catch!

Reply:
> Can total_joins be decremented? What happens if total_joins is decremented by more than the number of remaining_joins and then incremented again?

It can be decremented, yes. Originally it was going to be an API error to decrement `total_joins` by more than the current value of `remaining_joins` (to maintain `remaining_joins >= 0` as an invariant).

With `remaining_joins` removed, though, I think it's actually okay to allow `total_joins < join_count` - that just means future join attempts will fail until `total_joins` is appropriately incremented.

Reply:
I've gone ahead and removed `remaining_joins` in favor of `join_count`, like you suggested. I also did another pass on the doc comments to remove references to `remaining_joins` and improve clarity.